src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,242 @@
+            /// </summary>
+            /// <remarks>
+            /// <para>This call is not valid after <see cref="Dispose"/> is called. If this property or the target
+            /// object is used concurrently with a call to <see cref="Dispose"/>, it is possible for the code to be
+            /// using a disposed object. After the current instance is disposed, this property returns
+            /// <see langword="null"/>. However, the exact time when this property starts returning null after
+            /// <see cref="Dispose"/> is called is unspecified; code is expected to not use this property or the object
+            /// it returns after any code invokes <see cref="Dispose"/>.</para>
+            /// </remarks>
+            /// <value>The target object.</value>
+            public T Target => _instance;
+
+            /// <summary>
+            /// Increments the reference count for the disposable object, and returns a new disposable reference to it.
+            /// </summary>
+            /// <remarks>
+            /// <para>The returned object is an independent reference to the same underlying object. Disposing of the
+            /// returned value multiple times will only cause the reference count to be decreased once.</para>
+            /// </remarks>
+            /// <returns>A new <see cref="ReferenceCountedDisposable{T}"/> pointing to the same underlying object, if it
+            /// has not yet been disposed; otherwise, <see langword="null"/> if this reference to the underlying object
+            /// has already been disposed.</returns>
+            public ReferenceCountedDisposable<T> TryAddReference()
+            {
+                return TryAddReferenceImpl(_instance, _boxedReferenceCount);
+            }
+
+            /// <summary>
+            /// Provides the implementation for <see cref="TryAddReference"/> and
+            /// <see cref="WeakReference.TryAddReference"/>.
+            /// </summary>
+            private static ReferenceCountedDisposable<T> TryAddReferenceImpl(T target, StrongBox<int> referenceCount)
+            {
+                lock (referenceCount)
+                {
+                    if (referenceCount.Value == 0)
+                    {
+                        // The target is already disposed, and cannot be reused
+                        return null;
+                    }
+
+                    if (target == null)
+                    {
+                        // The current reference has been disposed, so even though it isn't disposed yet we don't have a
+                        // reference to the target
+                        return null;
+                    }
+
+                    referenceCount.Value++;
+
+                    // Must return a new instance, in order for the Dispose operation on each individual instance to
+                    // be idempotent.
+                    return new ReferenceCountedDisposable<T>(target, referenceCount);
```

Comment:
If you are returning new instances, it's unclear to me why the MemoryMappedInfo guy locks on this.